File: bench/src/spectralnorm/spectralnorm.cpp
#include<cmath>
#include<cstdio>
#include<cstdlib>
inline double eval_A(int i,int j){return 1.0/((double)((long long)(i+j)*(i+j+1)/2+i+1));}
void multiply_Av(const double*v,double*out,int n){for(int i=0;i<n;i++){double s=0.0;for(int j=0;j<n;j++)s+=v[j]*eval_A(i,j);out[i]=s;}}
void multiply_Atv(const double*v,double*out,int n){for(int i=0;i<n;i++){double s=0.0;for(int j=0;j<n;j++)s+=v[j]*eval_A(j,i);out[i]=s;}}
void multiply_AtAv(const double*v,double*out,int n){double*tmp=new double[n];multiply_Av(v,tmp,n);multiply_Atv(tmp,out,n);delete[] tmp;}
int main(int argc,char*argv[]){int n=argc>1?std::atoi(argv[1]):5500;double*u=new double[n];double*v=new double[n];for(int i=0;i<n;i++){u[i]=1.0;v[i]=0.0;}for(int k=0;k<10;k++){multiply_AtAv(u,v,n);multiply_AtAv(v,u,n);}double vBv=0.0,vv=0.0;for(int i=0;i<n;i++){vBv+=u[i]*v[i];vv+=v[i]*v[i];}std::printf("%.9f\n",std::sqrt(vBv/vv));delete[] u;delete[] v;return 0;}

File: bench/src/spectralnorm/spectralnorm.rs
use std::env;
fn eval_a(i: i64, j: i64) -> f64 {
    1.0 / (((i + j) * (i + j + 1) / 2 + i + 1) as f64)
}
fn multiply_av(v: &[f64], n: usize) -> Vec<f64> {
    let mut out = vec![0.0f64; n];
    for i in 0..n {
        let mut s = 0.0;
        for j in 0..n {
            s += v[j] * eval_a(i as i64, j as i64);
        }
        out[i] = s;
    }
    out
}
fn multiply_atv(v: &[f64], n: usize) -> Vec<f64> {
    let mut out = vec![0.0f64; n];
    for i in 0..n {
        let mut s = 0.0;
        for j in 0..n {
            s += v[j] * eval_a(j as i64, i as i64);
        }
        out[i] = s;
    }
    out
}
fn multiply_atav(v: &[f64], n: usize) -> Vec<f64> {
    multiply_atv(&multiply_av(v, n), n)
}
fn main() {
    let args: Vec<String> = env::args().collect();
    let n: usize = if args.len() > 1 {
        args[1].parse().unwrap()
    } else {
        5500
    };
    let mut u = vec![1.0f64; n];
    let mut v = vec![0.0f64; n];
    for _ in 0..10 {
        v = multiply_atav(&u, n);
        u = multiply_atav(&v, n);
    }
    let mut vbv = 0.0;
    let mut vv = 0.0;
    for i in 0..n {
        vbv += u[i] * v[i];
        vv += v[i] * v[i];
    }
    println!("{:.9}", (vbv / vv).sqrt());
}
